src/storage/data_directory.rs

```diff
@@ -0,0 +1,283 @@
+            None => Err(MSG_CID_NOT_FOUND),
+            Some(data) => Ok(data),
+        }
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event<T>() = default;
+
+        pub fn add_content(origin, data_object_type_id: <T as DOTRTrait>::DataObjectTypeId,
+                           id: T::ContentId, size: u64) {
+            // Origin has to be a member
+            let who = ensure_signed(origin).clone().unwrap();
+            if !T::IsActiveMember::is_active_member(&who) {
+                return Err(MSG_CREATOR_MUST_BE_MEMBER);
+            }
+
+            // Data object type has to be active
+            if !T::IsActiveDataObjectType::is_active_data_object_type(&data_object_type_id) {
+                return Err(MSG_DO_TYPE_MUST_BE_ACTIVE);
+            }
+
+            // We essentially accept the content ID and size at face value. All we
+            // need to know is that it doesn't yet exist.
+            let found = Self::content(&id);
+            ensure!(found.is_none(), MSG_DUPLICATE_CID);
+
+            // The liaison is something we need to take from staked roles. The idea
+            // is to select the liaison, for now randomly.
+            // FIXME without that module, we're currently hardcoding it, to the
+            // origin, which is wrong on many levels.
+            let liaison = who.clone();
+
+            // Let's create the entry then
+            let data: DataObject<T> = DataObject {
+                data_object_type: data_object_type_id,
+                size: size,
+                added_at_block: <system::Module<T>>::block_number(),
+                added_at_time: <timestamp::Module<T>>::now(),
+                origin: who,
+                liaison: liaison.clone(),
+                liaison_judgement: LiaisonJudgement::Pending,
+            };
+
+            // If we've constructed the data, we can store it and send an event.
+            <ContentMap<T>>::insert(id.clone(), data);
+            Self::deposit_event(RawEvent::ContentAdded(id, liaison));
+        }
+
+        // The LiaisonJudgement can be updated, but only by the liaison.
+        fn accept_content(origin, id: T::ContentId)
+        {
+            Self::update_content_judgement(origin, id.clone(), LiaisonJudgement::Accepted)?;
+            Self::deposit_event(RawEvent::ContentAccepted(id));
+        }
+
+        fn reject_content(origin, id: T::ContentId)
+        {
+            Self::update_content_judgement(origin, id.clone(), LiaisonJudgement::Rejected)?;
+            Self::deposit_event(RawEvent::ContentAccepted(id));
+        }
+    }
+}
+
+impl <T: Trait> Module<T> {
+    fn update_content_judgement(origin: T::Origin, id: T::ContentId, judgement: LiaisonJudgement) -> dispatch::Result
+    {
+        let who = ensure_signed(origin);
+
+        // Find the data
+        let found = Self::content(&id).ok_or(MSG_CID_NOT_FOUND);
```

Comment:
similar to prior point with regards to using `?;` pattern, so you don't need to unwrap.